src/System.Net.Utilities/src/System/Net/NetworkInformation/Ping.Unix.cs

```diff
@@ -2,33 +2,291 @@
+
+            Task finished = await Task.WhenAny(processCompletion.Task, timeoutTask).ConfigureAwait(false);
+            if (finished == timeoutTask && !p.HasExited)
+            {
+                // Try to kill the ping process if it didn't return. If it is already in the process of exiting, a Win32Exception will be thrown.
+                try
+                {
+                    p.Kill();
+                }
+                catch (Win32Exception) { }
+                throw new PingException(SR.net_ping_reply_timeout);
+            }
+            else
+            {
+                cts.Cancel();
+                if (p.ExitCode != 0)
+                {
+                    // This means no reply was received, although transmission may have been successful.
+                    throw new PingException(SR.net_ping_reply_timeout);
+                }
+
+                try
+                {
+                    string output = await p.StandardOutput.ReadToEndAsync().ConfigureAwait(false);
+                    int timeIndex = output.IndexOf("time=", StringComparison.Ordinal);
+                    int afterTime = timeIndex + "time=".Length;
+                    double parsedRtt;
+                    int msIndex = output.IndexOf("ms", afterTime);
+                    int numLength = msIndex - afterTime - 1;
+                    string timeSubstring = output.Substring(afterTime, numLength);
+                    parsedRtt = double.Parse(timeSubstring);
+                    long rtt = (long)Math.Round(parsedRtt);
+                    return new PingReply(
+                            address,
+                            null, // Ping utility cannot accomodate these, return null to indicate they were ignored.
+                            IPStatus.Success,
+                            rtt,
+                            Array.Empty<byte>()); // Ping utility doesn't deliver this info.
+                }
+                catch (Exception)
+                {
+                    // If the standard output cannot be successfully parsed, throw a generic PingException.
+                    throw new PingException(SR.net_ping);
+                }
+            }
+        }
+
         private void InternalDisposeCore()
         {
-            // TODO: Implement this (#2487)
+        }
+
+#if DEBUG
+        static Ping()
+        {
+            Debug.Assert(Marshal.SizeOf<IcmpHeader>() == 8, "The size of an ICMP Header must be 8 bytes.");
+        }
+#endif
+
+        // Must be 8 bytes total.
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct IcmpHeader
+        {
+            public byte Type;
+            public byte Code;
+            public ushort HeaderChecksum;
+            public ushort Identifier;
+            public ushort SequenceNumber;
+        }
+
+        private static unsafe byte[] CreateSendMessageBuffer(IcmpHeader header, byte[] payload)
+        {
+            var headerSize = Marshal.SizeOf<IcmpHeader>();
```

Comment:
Nits: s/var/int/, and IIRC you can just use `sizeof(IcmpHeader)` since you're in an unsafe context.